### PR TITLE
add notebook on the cause of miscalibration

### PR DESCRIPTION
**Pull request by glemaitre**
Notebook on the cause of miscalibration:

- Show examples in a 2D features space :
    - Mis-specified model: probabilistic XOR with non-zero probability for each class everywhere.
- Causes of miscalibration:
    - underfitting / mis-specified model (model class is not expressive enough, hyperparameters are too constraining, extra feature engineering is required for a given choice of the model class, regularization too strong, stopped the optimizer too early / failed to converge)
    - overfitting: not enough regularization (or too much)
    - use of class weights / undersampling in imbalanced classification problems.
    - covariate distribution shift

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Distribution shift is also a frequent cause of miscalibration

**Comment by glemaitre** [comment]
Thanks @GaelVaroquaux

**Comment by glemaitre** [comment]
> Distribution shift is also a frequent cause of miscalibration

@GaelVaroquaux One tiny question to get clearer insights: when it comes distribution shift, could we think of it similarly to the resampling effect meaning. As a concrete example, during a medical trial we learn on a controlled distribution where the proportion on the target is controlled in some way but when we are at predict time, then the class proportion is really different meaning that we are not calibrated anymore.

Is it this effect or you can have additional effect on the data `X` itself and have a feature distribution shift as well. In this case, I would expect that the model learnt is not valid anymore.

**Comment by GaelVaroquaux** [comment]
Actual covariate shift.

Taking the stance that the model is not valid anymore is fine from a theoretical point of view, but often you don't have enough data to learn a new model. With Alex Perez-Lebel, we found that often recalibrating gives you a significant fraction of the boost you could get with a new model (under review)

**Comment by glemaitre** [comment]
Thanks this is clear.

File: content/python_files/causes_miscalibration.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# %%
+import numpy as np
+
+
+def xor_generator(n_samples=1_000, seed=0):
+    rng = np.random.default_rng(seed)
+    X = rng.uniform(low=-3, high=3, size=(n_samples, 2))
+    y = np.logical_xor(X[:, 0] > 0, X[:, 1] > 0)
+    return X, y
```
Comment (ogrisel) [inline content/python_files/causes_miscalibration.py]: I think we should always use datasets with non-zero irreducible loss, even when synthetic: otherwise well-specified models can never be over-confident and we would not be able to illustrate the link between overfitting (e.g. lack of suitable regularization) and over confident calibration curves.

To generate a noisy XOR problem, one could use:

```python

import numpy as np


def xor_generator(n_samples=1_000, seed=0):
    rng = np.random.default_rng(seed)
    X = rng.uniform(low=-3, high=3, size=(n_samples, 2))
    noise = rng.normal(loc=0, scale=0.5, size=(n_samples, 2))
    y = np.logical_xor(X[:, 0] + noise[:, 0] > 0, X[:, 1] + noise[:, 1] > 0)
    return X, y
```

for instance.

File: content/python_files/causes_miscalibration.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from sklearn.preprocessing import SplineTransformer, PolynomialFeatures
+from sklearn.linear_model import LogisticRegression
+from sklearn.pipeline import make_pipeline
+
+model = make_pipeline(
+    SplineTransformer(n_knots=20),
+    PolynomialFeatures(degree=2, interaction_only=True),
+    LogisticRegression(),
+)
+model.fit(X, y)
```
Comment (ogrisel) [inline content/python_files/causes_miscalibration.py]: You should do a train test split with a small-ish training set (e.g. 1_000 observations) and a large enough test set (e.g. also 1_000 observations or more) to get a clean calibration curve if we don't do cross-validation.

Comment (ogrisel) [reply content/python_files/causes_miscalibration.py]: Then we could build different under fitting / over fitting profiles by changing the number of splines, including or removing polynomial interactions and changing regularization levels.

File: content/python_files/causes_miscalibration.py (hunk 1 above)
Comment (ogrisel) [reply content/python_files/causes_miscalibration.py]: the `noise` local variable could be renamed to `unobserved` to be more neutral. As you prefer.

**Comment by glemaitre** [comment]
Merging for the moment. We will make some PRs to improve the notebooks.
